### PR TITLE
Adding the alert info pagination API

### PR DESCRIPTION
**Pull request by qinxx108**
Implement the proposal here https://github.com/prometheus/alertmanager/issues/3361

### CODE REVIEW

File: api/v2/api_test.go [hunk 1]
```diff
@@ -510,3 +515,310 @@ receivers:
...
+			alertsProvider := newFakeAlerts(alerts)
+			api := API{
+				uptime:         time.Now(),
+				getAlertStatus: newGetAlertStatus(alertsProvider),
+				logger:         log.NewNopLogger(),
+				alerts:         alertsProvider,
+				setAlertStatus: func(model.LabelSet) {},
+			}
+			api.route = dispatch.NewRoute(&config.Route{Receiver: "def-receiver"}, nil)
+			r, err := http.NewRequest("GET", "/api/v2/alerts", nil)
```
Comment (emanlodovice) [inline api/v2/api_test.go:767]: Should this be `/api/v2/alertinfos`?

**Review by emanlodovice — CHANGES_REQUESTED** [review]

File: api/v2/restapi/operations/alertmanager_api.go [hunk 2]
```diff
@@ -366,7 +379,7 @@ func (o *AlertmanagerAPI) initHandlerCache() {
 	if o.handlers["POST"] == nil {
 		o.handlers["POST"] = make(map[string]http.Handler)
 	}
-	o.handlers["POST"]["/alerts"] = alert.NewPostAlerts(o.context, o.AlertPostAlertsHandler)
+	o.handlers["POST"]["/alertinfos"] = alert.NewPostAlerts(o.context, o.AlertPostAlertsHandler)
```
[the remark this answers is not in the document]
Comment (qinxx108) [reply api/v2/restapi/operations/alertmanager_api.go:382]: the openapi.yml get messed up. Fixed it

File: api/v2/restapi/operations/alert/post_alerts_urlbuilder.go [hunk 3]
```diff
@@ -49,7 +49,7 @@ func (o *PostAlertsURL) SetBasePath(bp string) {
 func (o *PostAlertsURL) Build() (*url.URL, error) {
 	var _result url.URL
 
-	var _path = "/alerts"
+	var _path = "/alertinfos"
```
[the remark this answers is not in the document]
Comment (qinxx108) [reply api/v2/restapi/operations/alert/post_alerts_urlbuilder.go:52]: the openapi.yml get messed up. Fixed it

File: api/v2/restapi/operations/alert/post_alerts.go [hunk 4]
```diff
@@ -44,7 +44,7 @@ func NewPostAlerts(ctx *middleware.Context, handler PostAlertsHandler) *PostAler
 }
 
 /*
-	PostAlerts swagger:route POST /alerts alert postAlerts
+	PostAlerts swagger:route POST /alertinfos alert postAlerts
```
[the remark this answers is not in the document]
Comment (qinxx108) [reply api/v2/restapi/operations/alert/post_alerts.go:47]: the openapi.yml get messed up. Fixed it

File: api/v2/client/alert/alert_client.go [hunk 5]
```diff
@@ -100,7 +100,7 @@ func (a *Client) PostAlerts(params *PostAlertsParams, opts ...ClientOption) (*Po
 	op := &runtime.ClientOperation{
 		ID:                 "postAlerts",
 		Method:             "POST",
-		PathPattern:        "/alerts",
+		PathPattern:        "/alertinfos",
```
[the remark this answers is not in the document]
Comment (qinxx108) [reply api/v2/client/alert/alert_client.go:103]: the openapi.yml get messed up. Fixed it

File: api/v2/client/alert/post_alerts_responses.go [hunk 6]
```diff
@@ -97,11 +97,11 @@ func (o *PostAlertsOK) IsCode(code int) bool {
 }
 
 func (o *PostAlertsOK) Error() string {
-	return fmt.Sprintf("[POST /alerts][%d] postAlertsOK ", 200)
+	return fmt.Sprintf("[POST /alertinfos][%d] postAlertsOK ", 200)
```
[the remark this answers is not in the document]
Comment (qinxx108) [reply api/v2/client/alert/post_alerts_responses.go:100]: the openapi.yml get messed up. Fixed it

File: api/v2/api.go [hunk 7]
```diff
@@ -721,3 +746,75 @@ func getSwaggerSpec() (*loads.Document, *analysis.Spec, error) {
...
+		if receiverFilter != nil && !receiversMatchFilter(receivers, receiverFilter) {
+			continue
+		}
+
+		if !alertFilter(a, now) {
+			continue
+		}
+
+		// Skip the alert if the next token is set and hasn't arrived the nextToken item yet.
+		if nextToken != nil && *nextToken >= a.Fingerprint().String() {
```
[the remark this answers is not in the document]
Comment (qinxx108) [reply api/v2/api.go:783]: done

File: api/v2/api.go [hunk 8]
```diff
@@ -259,50 +259,75 @@ func (api *API) getAlertsHandler(params alert_ops.GetAlertsParams) middleware.Re
...
+				fmt.Sprintf("failed to parse NextToken param: %v", *params.NextToken),
+			)
 	}
-	api.mtx.RUnlock()
 
+	alertFilter := api.alertFilter(matchers, *params.Silenced, *params.Inhibited, *params.Active)
+	alerts, err := api.getAlerts(ctx, receiverFilter, alertFilter, params.NextToken)
 	if err != nil {
 		level.Error(logger).Log("msg", "Failed to get alerts", "err", err)
 		return alert_ops.NewGetAlertsInternalServerError().WithPayload(err.Error())
```
[the remark this answers is not in the document]
Comment (qinxx108) [reply api/v2/api.go:320]: good catch, fixed

**Review by emanlodovice — APPROVED** [review]

**Pull request closed without merging**